system/puppet.py

```diff
@@ -101,6 +121,67 @@ def _write_structured_data(basedir, basename, data):
     out_file.write(json.dumps(data).encode('utf8'))
     out_file.close()
 
+class Rpmdatabase(object):
+    """
+    This is to capture rpm -qa
+    """
+
+    def __init__(self, module):
+        self.module            = module
+
+    def rpm(self):
+        # List installed packages and their version on fedora based systems
+        rpmlist = []
+        ts = rpm.TransactionSet()
+        mi = ts.dbMatch()
+        for h in mi:
+            rpmlist.append("%s-%s-%s.%s" % (h['name'], h['version'], h['release'],h['arch']))
+        return rpmlist
+
+class Debdatabase(object):
+    """
+    This is to capture rpm -qa equivelent on Debian based systems.
+    """
+
+    def __init__(self, module):
+        self.module            = module
+
+    def deb(self):
+        # List installed packages and their version on debian based systems
+        deblist = []
+        cache = apt.Cache()
+        for pkg in cache:
+            if pkg.candidate is not None and pkg.installed:
+                packagename =  pkg.name + '-' + pkg.candidate.version + '.' + pkg.candidate.architecture
+                deblist.append(packagename)
+        return deblist
+
+class Formatdiff(object):
+    """
+    This is to make the diff pretty
+    """
+
+    def __init__(self, module):
+        self.module            = module
+
+    def formatdiff(self, packagediff):
+        text = ''
+        seperator = '================================================================================\n'
+        diffheader = 'This begins the Package delta\n'
+        difffooter = 'The Package delta is over.\n'
+        # Only add lines begining with - or + immediately followed by a letter
+        for line in packagediff:
+            found = re.search("^[-|+][a-zA-Z].*", line)
+            if found:
+                text += line + '\n'
+        packagediff = seperator + diffheader + seperator + text + seperator + difffooter + seperator
+        return packagediff
+
+    def faildiff(self):
+        seperator = '================================================================================\n'
+        text = 'This OS does not support packagediff, please open a feature request on Github\n'
```

Comment:
The failure here is actually maybe two things- one that the OS is not supported - two that the OS is supported but python-apt/python-rpm isn't installed. It would be better to be able to differentiate so that the user could know that they could just apt-get install python-apt to get the feature.